extracted/konan.serializer/src/org/jetbrains/kotlin/descriptors/konan/KonanModuleDescriptorFactory.kt

```diff
@@ -0,0 +1,32 @@
+package org.jetbrains.kotlin.descriptors.konan
+
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.descriptors.impl.ModuleDescriptorImpl
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.storage.StorageManager
+
+interface KonanModuleDescriptorFactory {
+
+    /**
+     * Base method for creation of any Kotlin/Native [ModuleDescriptor].
+     */
+    fun createDescriptor(
+            name: Name,
+            storageManager: StorageManager,
+            builtIns: KotlinBuiltIns,
+            origin: KonanModuleOrigin,
+            customCapabilities: Map<ModuleDescriptor.Capability<*>, Any?> = emptyMap()
+    ): ModuleDescriptorImpl
+
+    /**
+     * Please use this method with care: As far as it creates an instance of [KotlinBuiltIns] it should be
```

Comment:
Is it possible to check that invariant in code?

Reply:
It is not possible right now. It could be possible if we would spend more time on further refactoring to forbid direct creation of individual `ModuleDescriptorImpl`s from individual `KonanLibrary`s.